libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1410,6 +1410,21 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 	else
 		_functionCall.annotation().type = make_shared<TupleType>(functionType->returnParameterTypes());
 
+	// Internal library functions must be implemented, because of inlining rules.
+	if (
+		functionType->kind() == FunctionType::Kind::Internal &&
+		functionType->hasDeclaration() &&
+		dynamic_cast<FunctionDefinition const*>(&functionType->declaration())
+	)
+	{
+		FunctionDefinition const* function = dynamic_cast<FunctionDefinition const*>(&functionType->declaration());
+		bool isLibraryFunction =
+			dynamic_cast<ContractDefinition const*>(function->scope()) &&
+			dynamic_cast<ContractDefinition const*>(function->scope())->isLibrary();
+		if (!function->isImplemented() && isLibraryFunction)
+			m_errorReporter.typeError(_functionCall.location(), "Inlined library function is lacking implementation.");
```

Comment:
Not sure if we should use the word "inlined" here. Isn't that an implementation detail of internal library functions?